multi_attribute_toydown/toydown.py

```diff
@@ -0,0 +1,340 @@
+                                (min, max) pairs for each attribute in each of the children, defining 
+                                the bounds on that attribute. Use None for one of min or max when there is 
+                                no bound in that direction.
+                                A value of "non-negative", flags that all counts should be > 0
+            parental_equality : Boolean flag - adds constraint that for all attributes, the sum of the 
+                                children's counts should be equal to that of the parent's counts.
+            maxiter           : Option to pass along to scipy optimizer.  Defines the maximum number of
+                                iterations the optimizer should preform.  Default value 200.
+            verbose           : Boolean flag -- if True prints debug info.
+        """
+
+        root = self.get_node(self.root)
+        
+        # noise tree
+        self.noise_tree(root)
+
+        if self.gurobi:
+            adj_root = self.adjusted_root_gp(root, None, None, verbose, self.pop_vap)
+        else:
+            if objective_fun == "L1": objective_fun = lambda n: lambda x: sp.linalg.norm(x-n, ord=1)
+            if objective_fun == "L2": objective_fun = lambda n: lambda x: sp.linalg.norm(x-n, ord=2)
+            adj_root = self.adjusted_root(root, objective_fun, node_cons, bounds,
+                                      maxiter, verbose)
+
+        adjusted = self.__adjust_tree(self, self.root, adj_root, objective_fun, node_cons, bounds,
+                           parental_equality, maxiter, verbose, self.gurobi, self.pop_vap)
+
+        if save: self.update_adjusted_and_error(adjusted)
+        return adjusted
+
+    def noise_tree(self, node):
+        self.add_laplacian_noise(node, self.eps_values[node.data.level])
+
+        if node.is_leaf():
+            return
+        
+        for child in self.children(node.identifier):
+            self.noise_tree(child)
+
+    @staticmethod
+    def __adjust_tree(model, node_id, node_adj, objective_fun, node_cons, bounds, 
+                                    parental_equality, maxiter, verbose, gurobi, pop_vap):
+        """ Recursively noises children and then "adjusts" the children to sum
+            up to the population of the parent.
+        """
+        if model.get_node(node_id).is_leaf():
+            return {node_id: node_adj}
+
+        # adjust children
+        if gurobi:
+             adj_children = ToyDown.adjust_children_gp(model.children(node_id), node_adj, 
+                                                            parental_equality, verbose, pop_vap)
+        else:
+            adj_children = ToyDown.adjust_children(model.children(node_id), node_adj, objective_fun, 
+                                               node_cons, bounds, parental_equality, maxiter, verbose)
+
+
+        # recurse
+        attribute_dict = {node_id: node_adj}
+        child_futures = []
```

Comment:
this variable seems like its not being used.

Reply:
Thanks!  Good catch